File: src/connection/quality.ts
/**
 * Connection quality scoring module.
 *
 * Takes raw connection test data (JEDEC reads, timing, status register)
 * and produces a 0-100 quality score with category breakdowns and
 * actionable diagnostic messages.
 */

// ── Types ──

export interface CategoryScore {
  name: string;
  score: number;     // 0-100
  weight: number;    // fraction (e.g. 0.50)
  diagnostic?: string;
}

export interface ConnectionQualityResult {
  score: number;           // 0-100 overall
  grade: string;           // "Excellent" | "Good" | "Fair" | "Poor"
  categories: CategoryScore[];
  diagnostics: string[];   // actionable recommendations
}

/** Raw data fed into the scoring engine. */
export interface RawConnectionData {
  jedecReadings: string[];       // hex strings per read, e.g. ["ef4017", "ef4017", ...]
  timingsMs: number[];           // per-read latency in ms
  statusRegisterOk: boolean;     // true if SR1 was readable
}

// ── Constants ──

const WEIGHT_CONSISTENCY   = 0.50;
const WEIGHT_JEDEC         = 0.20;
const WEIGHT_TIMING        = 0.15;
const WEIGHT_STATUS        = 0.15;

const INVALID_JEDEC = new Set(["000000", "ffffff"]);

// ── Helpers ──

function gradeFromScore(score: number): string {
  if (score >= 90) return "Excellent";
  if (score >= 70) return "Good";
  if (score >= 50) return "Fair";
  return "Poor";
}

/**
 * Compute consistency score (0-100).
 * Counts how many of the readings match the most common valid value.
 * JEDEC "000000" and "ffffff" are treated as failures, not valid samples.
 */
function scoreConsistency(readings: string[]): { score: number; diagnostic?: string } {
  // Filter out invalid JEDEC values
  const valid = readings.filter(r => !INVALID_JEDEC.has(r));
  if (valid.length === 0) {
    return {
      score: 0,
      diagnostic: "Reseat clip or probe — all JEDEC reads returned invalid data",
    };
  }

  // Most common valid value
  const freq = new Map<string, number>();
  for (const v of valid) {
    freq.set(v, (freq.get(v) ?? 0) + 1);
  }
  let bestCount = 0;
  for (const count of freq.values()) {
    if (count > bestCount) bestCount = count;
  }

  const ratio = bestCount / readings.length;
  const score = Math.round(ratio * 100);

  if (score < 70) {
    const bad = readings.length - bestCount;
    return {
      score,
      diagnostic: `Reseat SOIC clip — ${bad}/${readings.length} reads inconsistent`,
    };
  }
  return { score };
}

/**
 * Compute JEDEC validity score (0-100).
 * Checks whether the most-common reading is a valid (non-zero, non-FF) JEDEC ID.
 */
function scoreJedec(readings: string[]): { score: number; diagnostic?: string } {
  const valid = readings.filter(r => !INVALID_JEDEC.has(r));
  if (valid.length === 0) {
    return {
      score: 0,
      diagnostic: "Check chip orientation — JEDEC ID reads as all-zero or all-one",
    };
  }

  // All valid reads are good for JEDEC scoring
  const ratio = valid.length / readings.length;
  const score = Math.round(ratio * 100);

  if (score < 70) {
    return {
      score,
      diagnostic: "Check chip orientation — some JEDEC reads returned invalid values",
    };
  }
  return { score };
}

/**
 * Compute timing stability score (0-100).
 * Lower coefficient of variation = more stable = higher score.
 */
function scoreTiming(timings: number[]): { score: number; diagnostic?: string } {
  if (timings.length < 2) {
    return { score: 0, diagnostic: "Insufficient timing data to assess stability" };
  }

  const mean = timings.reduce((a, b) => a + b, 0) / timings.length;
  if (mean === 0) return { score: 100 };

  const variance = timings.reduce((sum, t) => sum + (t - mean) ** 2, 0) / timings.length;
  const stdDev = Math.sqrt(variance);
  const cv = stdDev / mean; // coefficient of variation

  // cv < 0.1 → perfect; cv > 1.0 → terrible
  const score = Math.round(Math.max(0, Math.min(100, (1 - cv) * 100)));

  if (score < 70) {
    return {
      score,
      diagnostic: "Shorten USB cable or remove hub — response times are unstable",
    };
  }
  return { score };
}

/**
 * Compute status register score (0 or 100).
 */
function scoreStatus(ok: boolean): { score: number; diagnostic?: string } {
  if (!ok) {
    return {
      score: 0,
      diagnostic: "Status register unreadable — verify VCC voltage and chip power",
    };
  }
  return { score: 100 };
}

// ── Main scoring function ──

/**
 * Compute an overall connection quality score from raw test data.
 *
 * Returns 0 if fewer than 2 successful reads are available
 * (not enough data for meaningful scoring).
 */
export function computeQualityScore(data: RawConnectionData): ConnectionQualityResult {
  // Guard: need at least 2 readings for meaningful stats
  if (data.jedecReadings.length < 2 || data.timingsMs.length < 2) {
    return {
      score: 0,
      grade: "Poor",
      categories: [
        { name: "Consistency",     score: 0, weight: WEIGHT_CONSISTENCY },
        { name: "JEDEC Validity",  score: 0, weight: WEIGHT_JEDEC },
        { name: "Timing Stability",score: 0, weight: WEIGHT_TIMING },
        { name: "Status Register", score: 0, weight: WEIGHT_STATUS },
      ],
      diagnostics: ["Not enough data — fewer than 2 successful reads"],
    };
  }

  const consistency = scoreConsistency(data.jedecReadings);
  const jedec       = scoreJedec(data.jedecReadings);
  const timing      = scoreTiming(data.timingsMs);
  const status      = scoreStatus(data.statusRegisterOk);

  const weighted =
    consistency.score * WEIGHT_CONSISTENCY +
    jedec.score       * WEIGHT_JEDEC +
    timing.score      * WEIGHT_TIMING +
    status.score      * WEIGHT_STATUS;

  const score = Math.round(weighted);
  const grade = gradeFromScore(score);

  const categories: CategoryScore[] = [
    { name: "Consistency",      score: consistency.score, weight: WEIGHT_CONSISTENCY, diagnostic: consistency.diagnostic },
    { name: "JEDEC Validity",   score: jedec.score,       weight: WEIGHT_JEDEC,       diagnostic: jedec.diagnostic },
    { name: "Timing Stability", score: timing.score,      weight: WEIGHT_TIMING,      diagnostic: timing.diagnostic },
    { name: "Status Register",  score: status.score,      weight: WEIGHT_STATUS,      diagnostic: status.diagnostic },
  ];

  const diagnostics: string[] = [];
  for (const cat of categories) {
    if (cat.diagnostic) diagnostics.push(cat.diagnostic);
  }

  return { score, grade, categories, diagnostics };
}

// ── Monitor helpers ──

const ANSI_GREEN  = "\x1b[32m";
const ANSI_RED    = "\x1b[31m";
const ANSI_YELLOW = "\x1b[33m";
const ANSI_RESET  = "\x1b[0m";

/** Threshold below which the monitor loop auto-exits. */
export const MONITOR_AUTO_EXIT_THRESHOLD = 20;

/**
 * Format a single monitor line showing the current score, trend arrow, and
 * optional degradation warning.
 *
 * @param currentScore  Current quality score (0-100)
 * @param previousScore Previous quality score (0-100), or null for first reading
 * @returns Formatted string with ANSI color codes
 */
export function formatMonitorLine(currentScore: number, previousScore: number | null): string {
  const grade = gradeFromScore(currentScore);
  const scoreStr = `${currentScore}/100 ${grade}`;

  // First reading — no trend
  if (previousScore === null) {
    return `Quality: ${scoreStr}`;
  }

  const delta = currentScore - previousScore;

  if (delta > 0) {
    // Improvement
    return `Quality: ${scoreStr} ${ANSI_GREEN}↑+${delta}${ANSI_RESET}`;
  } else if (delta < 0) {
    // Degradation
    const line = `Quality: ${scoreStr} ${ANSI_RED}↓${delta}${ANSI_RESET}`;
    if (currentScore < MONITOR_AUTO_EXIT_THRESHOLD) {
      return `${line} ${ANSI_RED}CRITICAL — score below ${MONITOR_AUTO_EXIT_THRESHOLD}, auto-exiting${ANSI_RESET}`;
    }
    if (Math.abs(delta) >= 15) {
      return `${line} ${ANSI_YELLOW}WARNING — significant degradation${ANSI_RESET}`;
    }
    return line;
  }

  // No change
  return `Quality: ${scoreStr} ↔ stable`;
}

/**
 * Returns true when the score is below the auto-exit threshold.
 * Separated from the monitor loop so it can be unit-tested.
 */
export function shouldAutoExit(score: number): boolean {
  return score < MONITOR_AUTO_EXIT_THRESHOLD;
}

File: rust/ratchet-core/src/connection/quality.rs
// Connection quality scoring  -  produces 0-100 quality score from raw JEDEC
// reads + timings + status-register flag. Ports src/connection/quality.ts.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CategoryScore {
    pub name: String,
    pub score: u32,
    pub weight: f32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub diagnostic: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ConnectionQualityResult {
    pub score: u32,
    pub grade: String,
    pub categories: Vec<CategoryScore>,
    pub diagnostics: Vec<String>,
}

#[derive(Debug, Clone)]
pub struct RawConnectionData {
    pub jedec_readings: Vec<String>,
    pub timings_ms: Vec<u32>,
    pub status_register_ok: bool,
}

const WEIGHT_CONSISTENCY: f32 = 0.50;
const WEIGHT_JEDEC: f32 = 0.20;
const WEIGHT_TIMING: f32 = 0.15;
const WEIGHT_STATUS: f32 = 0.15;

const INVALID_JEDEC: &[&str] = &["000000", "ffffff"];

pub fn grade_from_score(score: u32) -> &'static str {
    if score >= 90 {
        "Excellent"
    } else if score >= 70 {
        "Good"
    } else if score >= 50 {
        "Fair"
    } else {
        "Poor"
    }
}

fn is_invalid_jedec(s: &str) -> bool {
    INVALID_JEDEC.contains(&s)
}

fn score_consistency(readings: &[String]) -> (u32, Option<String>) {
    let valid: Vec<&String> = readings.iter().filter(|r| !is_invalid_jedec(r)).collect();
    if valid.is_empty() {
        return (
            0,
            Some("Reseat clip or probe  -  all JEDEC reads returned invalid data".to_string()),
        );
    }
    let mut freq: HashMap<&String, u32> = HashMap::new();
    for v in &valid {
        *freq.entry(*v).or_insert(0) += 1;
    }
    let best_count = freq.values().copied().max().unwrap_or(0);
    let ratio = best_count as f32 / readings.len() as f32;
    let score = (ratio * 100.0).round() as u32;
    if score < 70 {
        let bad = readings.len() as u32 - best_count;
        return (
            score,
            Some(format!(
                "Reseat SOIC clip  -  {bad}/{} reads inconsistent",
                readings.len()
            )),
        );
    }
    (score, None)
}

fn score_jedec(readings: &[String]) -> (u32, Option<String>) {
    let valid_count = readings.iter().filter(|r| !is_invalid_jedec(r)).count();
    if valid_count == 0 {
        return (
            0,
            Some("Check chip orientation  -  JEDEC ID reads as all-zero or all-one".to_string()),
        );
    }
    let ratio = valid_count as f32 / readings.len() as f32;
    let score = (ratio * 100.0).round() as u32;
    if score < 70 {
        return (
            score,
            Some("Check chip orientation  -  some JEDEC reads returned invalid values".to_string()),
        );
    }
    (score, None)
}

fn score_timing(timings: &[u32]) -> (u32, Option<String>) {
    if timings.len() < 2 {
        return (
            0,
            Some("Insufficient timing data to assess stability".to_string()),
        );
    }
    let mean: f64 = timings.iter().map(|t| *t as f64).sum::<f64>() / timings.len() as f64;
    if mean == 0.0 {
        return (100, None);
    }
    let variance: f64 = timings
        .iter()
        .map(|t| {
            let d = *t as f64 - mean;
            d * d
        })
        .sum::<f64>()
        / timings.len() as f64;
    let std_dev = variance.sqrt();
    let cv = std_dev / mean;
    let score = ((1.0 - cv) * 100.0).clamp(0.0, 100.0).round() as u32;
    if score < 70 {
        return (
            score,
            Some("Shorten USB cable or remove hub  -  response times are unstable".to_string()),
        );
    }
    (score, None)
}

fn score_status(ok: bool) -> (u32, Option<String>) {
    if ok {
        (100, None)
    } else {
        (
            0,
            Some("Status register unreadable  -  verify VCC voltage and chip power".to_string()),
        )
    }
}

pub fn compute_quality_score(data: &RawConnectionData) -> ConnectionQualityResult {
    if data.jedec_readings.len() < 2 || data.timings_ms.len() < 2 {
        return ConnectionQualityResult {
            score: 0,
            grade: "Poor".to_string(),
            categories: vec![
                CategoryScore {
                    name: "Consistency".to_string(),
                    score: 0,
                    weight: WEIGHT_CONSISTENCY,
                    diagnostic: None,
                },
                CategoryScore {
                    name: "JEDEC Validity".to_string(),
                    score: 0,
                    weight: WEIGHT_JEDEC,
                    diagnostic: None,
                },
                CategoryScore {
                    name: "Timing Stability".to_string(),
                    score: 0,
                    weight: WEIGHT_TIMING,
                    diagnostic: None,
                },
                CategoryScore {
                    name: "Status Register".to_string(),
                    score: 0,
                    weight: WEIGHT_STATUS,
                    diagnostic: None,
                },
            ],
            diagnostics: vec!["Not enough data  -  fewer than 2 successful reads".to_string()],
        };
    }

    let (consistency_s, consistency_d) = score_consistency(&data.jedec_readings);
    let (jedec_s, jedec_d) = score_jedec(&data.jedec_readings);
    let (timing_s, timing_d) = score_timing(&data.timings_ms);
    let (status_s, status_d) = score_status(data.status_register_ok);

    let weighted = consistency_s as f32 * WEIGHT_CONSISTENCY
        + jedec_s as f32 * WEIGHT_JEDEC
        + timing_s as f32 * WEIGHT_TIMING
        + status_s as f32 * WEIGHT_STATUS;
    let score = weighted.round() as u32;
    let grade = grade_from_score(score).to_string();

    let categories = vec![
        CategoryScore {
            name: "Consistency".to_string(),
            score: consistency_s,
            weight: WEIGHT_CONSISTENCY,
            diagnostic: consistency_d,
        },
        CategoryScore {
            name: "JEDEC Validity".to_string(),
            score: jedec_s,
            weight: WEIGHT_JEDEC,
            diagnostic: jedec_d,
        },
        CategoryScore {
            name: "Timing Stability".to_string(),
            score: timing_s,
            weight: WEIGHT_TIMING,
            diagnostic: timing_d,
        },
        CategoryScore {
            name: "Status Register".to_string(),
            score: status_s,
            weight: WEIGHT_STATUS,
            diagnostic: status_d,
        },
    ];
    let diagnostics: Vec<String> = categories
        .iter()
        .filter_map(|c| c.diagnostic.clone())
        .collect();

    ConnectionQualityResult {
        score,
        grade,
        categories,
        diagnostics,
    }
}

pub const MONITOR_AUTO_EXIT_THRESHOLD: u32 = 20;

pub fn format_monitor_line(current_score: u32, previous_score: Option<u32>) -> String {
    let grade = grade_from_score(current_score);
    let score_str = format!("{current_score}/100 {grade}");
    let Some(prev) = previous_score else {
        return format!("Quality: {score_str}");
    };
    let delta = current_score as i32 - prev as i32;
    if delta > 0 {
        format!("Quality: {score_str} \x1b[32m↑+{delta}\x1b[0m")
    } else if delta < 0 {
        let mut line = format!("Quality: {score_str} \x1b[31m↓{delta}\x1b[0m");
        if current_score < MONITOR_AUTO_EXIT_THRESHOLD {
            line.push_str(&format!(
                " \x1b[31mCRITICAL  -  score below {MONITOR_AUTO_EXIT_THRESHOLD}, auto-exiting\x1b[0m"
            ));
        } else if delta.abs() >= 15 {
            line.push_str(" \x1b[33mWARNING  -  significant degradation\x1b[0m");
        }
        line
    } else {
        format!("Quality: {score_str} ↔ stable")
    }
}

pub fn should_auto_exit(score: u32) -> bool {
    score < MONITOR_AUTO_EXIT_THRESHOLD
}

#[cfg(test)]
mod tests {
    use super::*;

    fn ten(s: &str) -> Vec<String> {
        vec![s.to_string(); 10]
    }

    #[test]
    fn perfect_run_yields_excellent_grade() {
        let data = RawConnectionData {
            jedec_readings: ten("ef4017"),
            timings_ms: vec![5; 10],
            status_register_ok: true,
        };
        let r = compute_quality_score(&data);
        assert!(r.score >= 90);
        assert_eq!(r.grade, "Excellent");
        assert!(r.diagnostics.is_empty());
    }

    #[test]
    fn all_invalid_jedec_is_poor() {
        let data = RawConnectionData {
            jedec_readings: ten("000000"),
            timings_ms: vec![5; 10],
            status_register_ok: true,
        };
        let r = compute_quality_score(&data);
        assert!(r.score < 50);
        assert_eq!(r.grade, "Poor");
        assert!(r.diagnostics.iter().any(|d| d.contains("Reseat")));
    }

    #[test]
    fn partial_consistency_drops_score() {
        let mut readings = ten("ef4017");
        readings[2] = "ab1234".to_string();
        readings[5] = "cd5678".to_string();
        readings[8] = "000000".to_string();
        let data = RawConnectionData {
            jedec_readings: readings,
            timings_ms: vec![5; 10],
            status_register_ok: true,
        };
        let r = compute_quality_score(&data);
        assert!(r.score < 90, "score should drop on partial consistency");
    }

    #[test]
    fn missing_status_register_subtracts_15() {
        let data = RawConnectionData {
            jedec_readings: ten("ef4017"),
            timings_ms: vec![5; 10],
            status_register_ok: false,
        };
        let r = compute_quality_score(&data);
        // Status weight = 0.15 → score drops by ~15 from perfect.
        assert!(r.score <= 85);
        assert!(r.diagnostics.iter().any(|d| d.contains("Status")));
    }

    #[test]
    fn unstable_timings_warn() {
        let data = RawConnectionData {
            jedec_readings: ten("ef4017"),
            timings_ms: vec![5, 500, 5, 700, 5, 800, 5, 900, 5, 1000],
            status_register_ok: true,
        };
        let r = compute_quality_score(&data);
        assert!(r.diagnostics.iter().any(|d| d.contains("response times")));
    }

    #[test]
    fn empty_data_returns_zero_score() {
        let data = RawConnectionData {
            jedec_readings: vec!["ef4017".to_string()],
            timings_ms: vec![5],
            status_register_ok: true,
        };
        let r = compute_quality_score(&data);
        assert_eq!(r.score, 0);
        assert_eq!(r.grade, "Poor");
    }

    #[test]
    fn grade_thresholds() {
        assert_eq!(grade_from_score(95), "Excellent");
        assert_eq!(grade_from_score(75), "Good");
        assert_eq!(grade_from_score(60), "Fair");
        assert_eq!(grade_from_score(30), "Poor");
    }

    #[test]
    fn monitor_first_reading_no_arrow() {
        let s = format_monitor_line(85, None);
        assert!(s.contains("85/100"));
        assert!(!s.contains("↑"));
        assert!(!s.contains("↓"));
    }

    #[test]
    fn monitor_improvement_shows_up_arrow() {
        let s = format_monitor_line(80, Some(60));
        assert!(s.contains("↑+20"));
    }

    #[test]
    fn monitor_critical_drop_warns() {
        let s = format_monitor_line(10, Some(80));
        assert!(s.contains("CRITICAL"));
    }

    #[test]
    fn monitor_significant_degradation_warns() {
        let s = format_monitor_line(50, Some(80));
        assert!(s.contains("WARNING"));
    }

    #[test]
    fn auto_exit_threshold() {
        assert!(should_auto_exit(10));
        assert!(!should_auto_exit(30));
    }
}
